### PR TITLE
8337167: StringSize deduplication

### PR DESCRIPTION
**Pull request by wenshao**
Integer.stringSize and Long.stringSize are used not only in java.lang, but also in other places. We put stringSize in DecimalDigits to reduce duplicate code and the use of JLA.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8337167](https://bugs.openjdk.org/browse/JDK-8337167): StringSize deduplication (**Enhancement** - P4)

### Reviewers
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - **Reviewer**)
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20322/head:pull/20322` \
`$ git checkout pull/20322`

Update a local copy of the PR: \
`$ git checkout pull/20322` \
`$ git pull https://git.openjdk.org/jdk.git pull/20322/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20322`

View PR using the GUI difftool: \
`$ git pr show -t 20322`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20322.diff">https://git.openjdk.org/jdk/pull/20322.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20322#issuecomment-2249503643)

### CODE REVIEW

File: src/java.base/share/classes/java/lang/Long.java [hunk 1]
```diff
@@ -499,18 +500,7 @@ public static String toUnsignedString(long i) {
      * code after loop unrolling.
      */
     static int stringSize(long x) {
```
Comment (liach) [inline src/java.base/share/classes/java/lang/Long.java:502]: Can we just remove this method?

File: src/java.base/share/classes/java/lang/Integer.java [hunk 2]
```diff
@@ -469,18 +470,7 @@ public static String toUnsignedString(int i) {
      * code after loop unrolling.
      */
     static int stringSize(int x) {
```
Comment (liach) [inline src/java.base/share/classes/java/lang/Integer.java:472]: Same question, if we can just inline this

File: src/java.base/share/classes/jdk/internal/access/JavaLangAccess.java [hunk 3]
```diff
@@ -459,8 +459,6 @@ public interface JavaLangAccess {
      */
     Object classData(Class<?> c);
 
-    int stringSize(long i);
```
Comment (liach) [inline src/java.base/share/classes/jdk/internal/access/JavaLangAccess.java:462]: :+1: Removing extra dependencies on `JavaLangAccess`

**Review by liach — APPROVED** [review]

**Comment by RogerRiggs** [comment]
What is the performance difference between the long and int versions?
If it is negligible, keep only the long version.

**Comment by liach** [comment]
@RogerRiggs They will be used in MethodHandle chains or explicit bytecode; having `long` and `int` specific versions simplify the bytecode or MethodHandle we need to prepare for String concatenation.

**Comment by wenshao** [comment]
@RogerRiggs The risk of removing stringSize(int) is not low, and the performance may be different on different platforms. If we want to remove it, it should be a separate PR, with extensive testing and extensive discussion before doing it.

**Review by RogerRiggs — APPROVED** [review]
lgtm

File: src/java.base/share/classes/jdk/internal/util/DecimalDigits.java [hunk 4]
```diff
@@ -110,4 +110,30 @@ public static int stringSize(int x) {
         }
         return 10 + d;
     }
+
+    /**
+     * Returns the string representation size for a given long value.
```
Comment (turbanoff) [inline src/java.base/share/classes/jdk/internal/util/DecimalDigits.java:115]: But it returns not a _string representation_.

**Pull request closed without merging**